### PR TITLE
Fix `podman system reset` with external containers

### PR DESCRIPTION
**Pull request by mheon**
It looks like we had some logic for this from #10789 but it does not appear to have ever worked; we can't pull external containers out of the DB, so the ContainerRm call failed unconditionally.

Instead, just handle them in Libpod when we're removing images. We're removing every image, so setting Force when removing images should get rid of all external containers. It's a little later in the process than the current (nonfunctional) solution is but I can't think of a reason why that would be bad.

[NO NEW TESTS NEEDED] We do not currently test `system reset`. We should probably reevaluate that at some point this year.

Fixes https://issues.redhat.com/browse/RHEL-21261

#### Does this PR introduce a user-facing change?

```release-note
Fixed a bug where `podman system reset -f` would fail when non-Podman containers were present.
```

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
@mheon could you add a card to the backlog for system reset tests please?

Code LGTM, @Luap99 PTAL

**Comment by martinpitt** [comment]
I'm a bit stumped by the failures here -- [cockpit-podman rawhide failed](https://artifacts.dev.testing-farm.io/01ce7d57-2f12-4702-a993-b3ce79722616/) like this:
```
# podman run -d -p 6000:5000 --name registry_alt --stop-timeout 0 localhost/test
Error: cannot listen on the TCP port: listen tcp4 :6000: bind: address already in use
```

and similiarly on [f39](https://artifacts.dev.testing-farm.io/291ce325-18af-41ac-9f2a-345fc254f1f4/):
```
# podman run -d --pod testpod1 --name test-pod-1 --stop-timeout 0 localhost/test-alpine sleep 500
time="2024-01-10T21:22:22Z" level=error msg="Starting some container dependencies"
time="2024-01-10T21:22:22Z" level=error msg="\"rootlessport listen tcp 0.0.0.0:6000: bind: address already in use\""
Error: starting some containers: internal libpod error
```

where `ps -a` is as expected:
```
CONTAINER ID  IMAGE                                                  COMMAND     CREATED        STATUS      PORTS                                                                        NAMES
ec0e9a595ac3  localhost/podman-pause:5.0.0-dev-dd1330a4e-1704920869              3 seconds ago  Created     0.0.0.0:6000->5000/tcp, 127.0.0.1:6001->5001/udp, 127.0.0.2:44925->9001/tcp  5579bc451063-infra
e40ed81a2eaf  localhost/test-alpine:latest                           sleep 500   3 seconds ago  Created     0.0.0.0:6000->5000/tcp, 127.0.0.1:6001->5001/udp, 127.0.0.2:44925->9001/tcp  test-pod-1
```

c-podman doesn't test `system reset` either, so unrelated to *this* PR. Of course you'll need at least one more push to fix the [arm static check failure](https://github.com/containers/podman/pull/21226/checks?check_run_id=20361034499), so we'll get a retry here.

I haven't seen this failure mode before, and recent podman PRs were also green. In the last 12 hours we got a [netavark](https://copr.fedorainfracloud.org/coprs/rhcontainerbot/podman-next/build/6880057/) build in podman-next. But this is a race condition, it doesn't happen every time, and I can't immediately reproduce it. I'll stare at the log for a little longer, maybe I'll get an idea.

Locally I can at least reproduce the `testHealthcheck` failure. That is our fault, I sent a fix to https://github.com/cockpit-project/cockpit-podman/pull/1533

**Comment by Luap99** [comment]
> I'm a bit stumped by the failures here -- [cockpit-podman rawhide failed](https://artifacts.dev.testing-farm.io/01ce7d57-2f12-4702-a993-b3ce79722616/) like this:
> 
> ```
> # podman run -d -p 6000:5000 --name registry_alt --stop-timeout 0 localhost/test
> Error: cannot listen on the TCP port: listen tcp4 :6000: bind: address already in use
> ```
> 
> and similiarly on [f39](https://artifacts.dev.testing-farm.io/291ce325-18af-41ac-9f2a-345fc254f1f4/):
> 
> ```
> # podman run -d --pod testpod1 --name test-pod-1 --stop-timeout 0 localhost/test-alpine sleep 500
> time="2024-01-10T21:22:22Z" level=error msg="Starting some container dependencies"
> time="2024-01-10T21:22:22Z" level=error msg="\"rootlessport listen tcp 0.0.0.0:6000: bind: address already in use\""
> Error: starting some containers: internal libpod error
> ```
> 
> where `ps -a` is as expected:
> 
> ```
> CONTAINER ID  IMAGE                                                  COMMAND     CREATED        STATUS      PORTS                                                                        NAMES
> ec0e9a595ac3  localhost/podman-pause:5.0.0-dev-dd1330a4e-1704920869              3 seconds ago  Created     0.0.0.0:6000->5000/tcp, 127.0.0.1:6001->5001/udp, 127.0.0.2:44925->9001/tcp  5579bc451063-infra
> e40ed81a2eaf  localhost/test-alpine:latest                           sleep 500   3 seconds ago  Created     0.0.0.0:6000->5000/tcp, 127.0.0.1:6001->5001/udp, 127.0.0.2:44925->9001/tcp  test-pod-1
> ```
> 
> c-podman doesn't test `system reset` either, so unrelated to _this_ PR. Of course you'll need at least one more push to fix the [arm static check failure](https://github.com/containers/podman/pull/21226/checks?check_run_id=20361034499), so we'll get a retry here.
> 
> I haven't seen this failure mode before, and recent podman PRs were also green. In the last 12 hours we got a [netavark](https://copr.fedorainfracloud.org/coprs/rhcontainerbot/podman-next/build/6880057/) build in podman-next. But this is a race condition, it doesn't happen every time, and I can't immediately reproduce it. I'll stare at the log for a little longer, maybe I'll get an idea.
> 
> Locally I can at least reproduce the `testHealthcheck` failure. That is our fault, I sent a fix to [cockpit-project/cockpit-podman#1533](https://github.com/cockpit-project/cockpit-podman/pull/1533)

Did they got the a new kernel? We had several issues with kernel regressions in regards to the bind() syscall in the past.
Netavark should not matter at all for such an error as it does not bind any ports and I certainly am not aware of any podman changes that touched any of this code

**Review by Luap99 — COMMENTED** [review]
code LGTM but we do have system reset tests, see test/e2e/system_reset_test.go

**Comment by martinpitt** [comment]
@Luap99 Not [in the last few days](https://bodhi.fedoraproject.org/updates/?packages=kernel); 4 hours ago there was a new upload, but it's not yet in -updates. I'm testing locally against 6.6.9-200.fc39.x86_64 , TF ran 6.6.8-200. [This log](https://artifacts.dev.testing-farm.io/291ce325-18af-41ac-9f2a-345fc254f1f4/) certainly sounds like some initial test does something that keeps port 6000 busy in the "background" despite killing all containers, pods, etc, and then breaks any subsequent test which tries to publish port 6000. Unfortunately the [journal](https://artifacts.dev.testing-farm.io/291ce325-18af-41ac-9f2a-345fc254f1f4/work-podman-usermla2alfz/plans/cockpit-podman/podman-user/execute/data/guest/default-0/test/browser/user-1/data/TestApplication-testCreatePodUser-fedora-39-127.0.0.1-22-FAIL.log.gz) doesn't have any useful error message for that, it just shows the previous accesses to the "localhost:6000" registry which `TestApplication.testBasicUser` (first test that ran) started. At the end of each test, all container get stopped (`podman rm --all --force`, same with `pod`) and /var/lib/containers / ~/.local/share/containers gets restored to the pristine default. So somehow port 6000 survives that.

Well, l'll keep watching for this. If it happens again, I'll throw some more time on it and try to repro on a real TF instance.

**Comment by martinpitt** [comment]
@Luap99 No, we are not running system reset, what I [mentioned above](https://github.com/containers/podman/pull/21226#issuecomment-1886635046). There is certainly *a* bug here, but not in this PR in particular. But note  [this failure](https://github.com/containers/podman/pull/21226/checks?check_run_id=20361034499), which does look related to this PR.

**Comment by Luap99** [comment]
How sure are you because I did a quick search I see this: https://github.com/cockpit-project/cockpit-podman/blob/31a410f1a8ec89d3bc5d5b9f4d6e66b9c9d80b0d/test/check-application#L116-L123?

Assuming a prior container bound the port it can only be released if the conmon process died as well, so the root cause is likely leaking conmon somewhere.

**Comment by Luap99** [comment]
Ok this is trivial to reproduce with this PR locally,
```
$ bin/podman run -d -p 8080:80  alpine sleep inf
$ bin/podman --log-level debug system reset --force
$ bin/podman ps -a
CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
$ ps aux | grep conmon
pholzing   17043  0.0  0.0 225932  2064 ?        Ss   11:35   0:00 /usr/bin/conmon --api-versio...
```

Thus this path is clearly not working, for some reason we are able to delete containers without stopping them first which hints at a much bigger issue in the reset code somewhere.

**Comment by mheon** [comment]
The container is gone (was successfully unmounted and removed), but `sleep inf` is still happily running. Fascinating.

**Comment by mheon** [comment]
Repushed with a very large commit that refactors the way `system renumber`, `system reset`, and `system migrate` work. Cleans up the code a lot and fixes the bug.

Will add two tests to verify both our broken cases.

**Comment by mheon** [comment]
One test added. I don't think we can easily get Buildah into the E2E tests to create a non-Podman container and test the original functionality of this PR.

File: test/e2e/system_reset_test.go [hunk 1]
```diff
@@ -74,4 +76,28 @@ var _ = Describe("podman system reset", Serial, func() {
...
+		port2 := GetPort()
+		session := podmanTest.Podman([]string{"run", "--name", ctrName, "-p", fmt.Sprintf("%d:%d", port1, port2), "-d", ALPINE, "top"})
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(ExitCleanly())
+
+		reset := podmanTest.Podman([]string{"system", "reset", "--force"})
+		reset.WaitWithDefaultTimeout()
+		Expect(reset).Should(ExitCleanly())
+
+		session2 := podmanTest.Podman([]string{"run", "--name", ctrName, "-p", fmt.Sprintf("%d:%d", port1, port2), "-d", ALPINE, "top"})
```
Comment (Luap99) [inline test/e2e/system_reset_test.go:99]: you may want to add a comment on the importance of -p here.

**Review by Luap99 — APPROVED** [review]
changes LGTM
Although cockpit tests still seem to fail but I cannot read that log so I don't know if this is related so I just assume its not.

**Comment by martinpitt** [comment]
@Luap: Just to make sure, when you said "I cannot read that log", you meant you can't interpret https://artifacts.dev.testing-farm.io/c495aa66-9ba6-435d-b2c4-da2942e67b98/ , not that you can't access it? It's public without any login.

Anyway, not your or our fault, someone just broke firefox in rawhide :crying_cat_face: 
```
XPCOMGlueLoad error for file /usr/lib64/firefox/libmozgtk.so:
/lib64/libgdk_pixbuf-2.0.so.0: undefined symbol: g_once_init_leave_pointer
```

Until then, *all* rawhide cockpit logs will be broken. Please ignore.

@jelly can you please have a look on Monday? this at least needs an urgent bugzilla and yelling at some bodhi update -- this is so obviously broken.. (I'm on PTO until Thu)

**Comment by mheon** [comment]
OK, we should be good to merge, then. @containers/podman-maintainers PTAL

**Pull request merged**